--- a/Makefile.tail	2020-09-06 21:02:34.659855219 +0000
+++ b/Makefile.tail	2020-09-06 21:03:16.661259983 +0000
@@ -23,9 +23,14 @@
 install:
 	cd $(LIVEMEDIA_DIR) ; $(MAKE) install
 	cd $(GROUPSOCK_DIR) ; $(MAKE) install
+	cd $(MEDIA_SERVER_DIR) ; $(MAKE) install
+	cd $(PROXY_SERVER_DIR) ; $(MAKE) install
+
+installstaging:
+	cd $(LIVEMEDIA_DIR) ; $(MAKE) install
+	cd $(GROUPSOCK_DIR) ; $(MAKE) install
 	cd $(USAGE_ENVIRONMENT_DIR) ; $(MAKE) install
 	cd $(BASIC_USAGE_ENVIRONMENT_DIR) ; $(MAKE) install
-	cd $(TESTPROGS_DIR) ; $(MAKE) install
 	cd $(MEDIA_SERVER_DIR) ; $(MAKE) install
 	cd $(PROXY_SERVER_DIR) ; $(MAKE) install
 
