--- kernel-3.10.14/arch/mips/xburst/soc-t20/chip-t20/isvp/common/spi_bus.c
+++ kernel-3.10.14/arch/mips/xburst/soc-t20/chip-t20/isvp/common/spi_bus.c
@@ -114,6 +114,27 @@ struct spi_nor_platform_data spi_nor_pdata[] = {
 #endif
 	},
 	{
+		.name           = "EN25QH128A",
+		.pagesize       = 256,
+		.sectorsize     = 4 * 1024,
+		.chipsize       = 16384 * 1024,
+		.erasesize      = 32 * 1024,
+		.id             = 0x1c7018,
+
+		.block_info     = flash_block_info,
+		.num_block_info = ARRAY_SIZE(flash_block_info),
+
+		.addrsize       = 3,
+		.pp_maxbusy     = 3,            /* 3ms */
+		.se_maxbusy     = 400,          /* 400ms */
+		.ce_maxbusy     = 8 * 10000,    /* 80s */
+
+		.st_regnum      = 3,
+#ifdef CONFIG_SPI_QUAD
+		.quad_mode = &flash_quad_mode[0],
+#endif
+	},
+	{
 		.name           = "EN25QH64",
 		.pagesize       = 256,
 		.sectorsize     = 4 * 1024,
@@ -407,6 +428,29 @@ struct spi_nor_platform_data spi_nor_pdata[] = {
 		.quad_mode = &flash_quad_mode[0],
 #endif
 	},
+	{
+		.name           = "ZB25VQ128",
+		.pagesize       = 256,
+		.sectorsize     = 4 * 1024,
+		.chipsize       = 16384 * 1024,
+		.erasesize      = 32 * 1024,
+		.id             = 0x5e4018,
+
+		.block_info     = flash_block_info,
+		.num_block_info = ARRAY_SIZE(flash_block_info),
+
+		.addrsize       = 3,
+		.pp_maxbusy     = 3,            /* 3ms */
+		.se_maxbusy     = 400,          /* 400ms */
+		.ce_maxbusy     = 8 * 10000,    /* 80s */
+
+		.st_regnum      = 3,
+#ifdef CONFIG_SPI_QUAD
+		.quad_mode = &flash_quad_mode[0],
+#endif
+	},
+
+
 };
 
 struct jz_sfc_info sfc_info_cfg = {
