diff -Nura kernel-3.10.14/arch/mips/xburst/soc-t20/chip-t20/isvp/common/board_base.c kernel-3.10.14_led/arch/mips/xburst/soc-t20/chip-t20/isvp/common/board_base.c
--- kernel-3.10.14/arch/mips/xburst/soc-t20/chip-t20/isvp/common/board_base.c	2018-09-21 16:41:27.000000000 +0200
+++ kernel-3.10.14_led/arch/mips/xburst/soc-t20/chip-t20/isvp/common/board_base.c	2018-11-12 02:44:38.462211000 +0100
@@ -28,6 +28,9 @@
 #ifdef CONFIG_KEYBOARD_GPIO
 	DEF_DEVICE(&jz_button_device, 0, 0),
 #endif
+#if defined(CONFIG_LEDS_GPIO) || defined(CONFIG_LEDS_GPIO_MODULE)
+	DEF_DEVICE(&jz_led_device, 0, 0),
+#endif
 #ifdef CONFIG_I2C_GPIO
 #ifdef CONFIG_SOFT_I2C0_GPIO_V12_JZ
 	DEF_DEVICE(&i2c0_gpio_device, 0, 0),
diff -Nura kernel-3.10.14/arch/mips/xburst/soc-t20/chip-t20/isvp/common/board_base.h kernel-3.10.14_led/arch/mips/xburst/soc-t20/chip-t20/isvp/common/board_base.h
--- kernel-3.10.14/arch/mips/xburst/soc-t20/chip-t20/isvp/common/board_base.h	2018-09-21 16:41:27.000000000 +0200
+++ kernel-3.10.14_led/arch/mips/xburst/soc-t20/chip-t20/isvp/common/board_base.h	2018-11-12 02:44:52.814331000 +0100
@@ -9,6 +9,9 @@
 #ifdef CONFIG_KEYBOARD_GPIO
 extern struct platform_device jz_button_device;
 #endif
+#if defined(CONFIG_LEDS_GPIO) || defined(CONFIG_LEDS_GPIO_MODULE)
+extern struct platform_device jz_led_device;
+#endif
 #ifdef CONFIG_INV_MPU_IIO
 extern struct mpu_platform_data mpu9250_platform_data;
 #endif
diff -Nura kernel-3.10.14/arch/mips/xburst/soc-t20/chip-t20/isvp/common/led_gpio.c kernel-3.10.14_led/arch/mips/xburst/soc-t20/chip-t20/isvp/common/led_gpio.c
--- kernel-3.10.14/arch/mips/xburst/soc-t20/chip-t20/isvp/common/led_gpio.c	1970-01-01 01:00:00.000000000 +0100
+++ kernel-3.10.14_led/arch/mips/xburst/soc-t20/chip-t20/isvp/common/led_gpio.c	2018-11-12 04:56:52.771282964 +0100
@@ -0,0 +1,46 @@
+#include <linux/platform_device.h>
+#include <linux/leds.h>
+#include "board_base.h"
+
+/* GPIO LED */
+#if defined(CONFIG_LEDS_GPIO) || defined(CONFIG_LEDS_GPIO_MODULE)
+struct gpio_led board_led[] = {
+	{
+		.name = "led_blue",
+		.default_trigger = "mmc0",
+		.gpio = 39,
+		.active_low = 1,
+	},
+	{
+		.name = "led_yellow",
+		.default_trigger = "timer",
+		.gpio = 38,
+		.active_low = 1,
+	},
+	{
+		.name = "led_blue1",
+		.default_trigger = "mmc0",
+		.gpio = 76,
+		.active_low = 1,
+	},
+	{
+		.name = "led_yellow1",
+		.default_trigger = "timer",
+		.gpio = 75,
+		.active_low = 1,
+	},
+};
+
+static struct gpio_led_platform_data board_led_data = {
+	.num_leds	= ARRAY_SIZE(board_led),
+	.leds		= board_led
+};
+
+struct platform_device jz_led_device = {
+	.name		= "leds-gpio",
+	.id		= -1,
+	.dev		= {
+	.platform_data	= &board_led_data,
+	},
+};
+#endif /* CONFIG_LEDS_GPIO */
diff -Nura kernel-3.10.14/arch/mips/xburst/soc-t20/chip-t20/isvp/common/Makefile kernel-3.10.14_led/arch/mips/xburst/soc-t20/chip-t20/isvp/common/Makefile
--- kernel-3.10.14/arch/mips/xburst/soc-t20/chip-t20/isvp/common/Makefile	2018-09-21 16:41:27.000000000 +0200
+++ kernel-3.10.14_led/arch/mips/xburst/soc-t20/chip-t20/isvp/common/Makefile	2018-11-12 02:16:33.209290337 +0100
@@ -2,6 +2,7 @@
 obj-y		+= board_base.o
 obj-y		+= misc.o
 obj-y		+= mmc.o
+obj-y		+= led_gpio.o
 obj-$(CONFIG_KEYBOARD_GPIO)	+= keyboard_gpio.o
 
 obj-$(CONFIG_BCM_PM_CORE)   += bcm_power_control.o
@@ -9,4 +10,4 @@
 obj-$(CONFIG_I2C_GPIO)			+= i2c_bus.o
 obj-$(CONFIG_I2C0_V12_JZ)		+= i2c_bus.o
 obj-$(CONFIG_I2C1_V12_JZ)		+= i2c_bus.o
-obj-$(CONFIG_MTD_JZ_SFC_NORFLASH)	+= spi_bus.o
\ No newline at end of file
+obj-$(CONFIG_MTD_JZ_SFC_NORFLASH)	+= spi_bus.o
