--- uboot/drivers/spi/jz_spi.h	2018-09-28 03:16:08.000000000 +0200
+++ jz_spi.h	2019-03-22 18:13:21.298601800 +0100
@@ -349,6 +349,27 @@
 		},
 	},
 	{
+		.name           = "EN25QH128A",
+		.id_manufactory = 0x1c7018,
+		.page_size       = 256,
+		.sector_size      = (32 * 1024),
+		.addr_size = 3,
+		.size           = (16 * 1024 * 1024),
+		.quad_mode = {
+			.dummy_byte = 8,
+			.RDSR_CMD = CMD_RDSR_1,
+			.WRSR_CMD = CMD_WRSR_1,
+			.RDSR_DATE = 0x2,//the data is write the spi status register for QE bit
+			.RD_DATE_SIZE = 1,
+			.WRSR_DATE = 0x2,//this bit should be the flash QUAD mode enable
+			.WD_DATE_SIZE = 1,
+			.cmd_read = CMD_QUAD_READ,
+#ifdef CONFIG_JZ_SFC
+			.sfc_mode = TRAN_SPI_QUAD,
+#endif
+		},
+	},
+	{
 		.name           = "FM25Q64",
 		.id_manufactory = 0xF83217,
 		.page_size       = 256,
